mm2src/coins/nft.rs

```diff
@@ -306,13 +292,60 @@ pub async fn update_nft(ctx: MmArc, req: UpdateNftReq) -> MmResult<(), UpdateNft
             &req.url_antispam,
         )
         .await?;
+        update_nft_global_in_coins_ctx(&ctx, &storage, *chain, nft_global).await?;
         update_transfers_with_empty_meta(&storage, chain, &req.url, &req.url_antispam).await?;
         update_spam(&storage, *chain, &req.url_antispam).await?;
         update_phishing(&storage, chain, &req.url_antispam).await?;
     }
     Ok(())
 }
 
+/// Updates the global NFT information in the coins context.
+///
+/// This function uses the up to date NFT list for a given chain and updates the
+/// corresponding global NFT information in the coins context.
+async fn update_nft_global_in_coins_ctx<T>(
+    ctx: &MmArc,
+    storage: &T,
+    chain: Chain,
+    mut nft_global: EthCoin,
+) -> MmResult<(), UpdateNftError>
+where
+    T: NftListStorageOps + NftTransferHistoryStorageOps,
+{
+    let nft_list = storage.get_nft_list(vec![chain], true, 1, None, None).await?;
+    let coins_ctx = CoinsContext::from_ctx(ctx).map_to_mm(UpdateNftError::Internal)?;
+    let mut coins = coins_ctx.coins.lock().await;
+
+    update_nft_infos(&mut nft_global, nft_list.nfts).await;
+
+    coins.insert(
+        nft_global.ticker.to_string(),
+        MmCoinStruct::new(MmCoinEnum::EthCoin(nft_global)),
+    );
+    Ok(())
+}
+
+async fn update_nft_infos(nft_global: &mut EthCoin, nft_list: Vec<Nft>) {
+    let new_nft_infos: HashMap<String, NftInfo> = nft_list
+        .into_iter()
+        .map(|nft| {
+            let key = format!("{},{}", nft.common.token_address, nft.token_id);
+            let nft_info = NftInfo {
+                token_address: nft.common.token_address,
+                token_id: nft.token_id,
+                chain: nft.chain,
+                contract_type: nft.contract_type,
+                amount: nft.common.amount,
+            };
+            (key, nft_info)
+        })
+        .collect();
+
+    let mut global_nft_infos = nft_global.nfts_infos.lock().await;
+    *global_nft_infos = new_nft_infos;
```

Comment:
This seems scary. We read some data from the storage and replace the mutexed data without any sanity control if we lost some data or not. I don't know if it's possible optimizing this entirely, but in worst case scenario I would at least upsert the new data instead of replacing the old/existing data entirely.

Reply:
I understand your feelings, but we kinda need to "loose" some data, bcz if user transferred erc721 or all erc1155 nfts, they are no longer have it in wallet. 

What we do is update all nft info: all nfts owned by user and their transfer history. Then we use up-to-date database info to update `nfts_infos` field in global nft in `CoinsContext`. If user transferred some nft, he no longer is owner, so we wont have it in storage => we dont need it in `CoinsContext`.

instead of `*global_nft_infos = new_nft_infos` we can do this
```rust
    let mut global_nft_infos = nft_global.nfts_infos.lock().await;
    // Remove keys that are not present in new_nft_infos
    global_nft_infos.retain(|key, _| new_nft_infos.contains_key(key));

    // Upsert the new_nft_infos into global_nft_infos
    for (key, new_info) in new_nft_infos {
        global_nft_infos
            .entry(key)
            .and_modify(|existing_info| *existing_info = new_info.clone())
            .or_insert(new_info);
    }
```
In both implementations we lock `nfts_infos` field to prevent race conditions. 
As a result both approaches do the same, just `retain` and for loop operations will take more time then this `*global_nft_infos = new_nft_infos`

I dont insist on current impl, but Im not sure that manual update worth it.
So what do you think?

Reply:
Thanks for the explanation, could you please add some safety notes (like you explained here) on the updating line?

Reply:
Sure, done